src/biokbase/narrative/jobs/job.py

```diff
@@ -77,65 +117,104 @@ def __init__(self, **kwargs):
+            ),
+            batch_job=lambda: self._const_state.get("batch_job", JOB_ATTR_DEFAULTS["batch_job"]),
+            cell_id=lambda: self._const_state.get("job_input", {}).get("narrative_cell_info", {}).get("cell_id", JOB_ATTR_DEFAULTS["cell_id"]),
+            child_jobs=lambda: self.state().get("child_jobs", JOB_ATTR_DEFAULTS["child_jobs"]),
+            job_id=lambda: self._const_state.get("job_id"),
+            params=lambda: self._const_state.get("job_input", {}).get("params", JOB_ATTR_DEFAULTS["params"]),
+            retry_ids=lambda: self.state().get("retry_ids", JOB_ATTR_DEFAULTS["retry_ids"]),
+            retry_parent=lambda: self._const_state.get("retry_parent", JOB_ATTR_DEFAULTS["retry_parent"]),
+            run_id=lambda: self._const_state.get("job_input", {}).get("narrative_cell_info", {}).get("run_id", JOB_ATTR_DEFAULTS["run_id"]),
+            tag=lambda: self._const_state.get("job_input", {}).get("narrative_cell_info", {}).get("tag", JOB_ATTR_DEFAULTS["tag"]),
+            user=lambda: self._const_state.get("user", JOB_ATTR_DEFAULTS["user"]),
         )
 
+        if name not in attr:
+            raise AttributeError(f"'Job' object has no attribute '{name}'")
+
+        return attr[name]()
+
+    def __setattr__(self, name, value):
+        if name in STATE_ATTRS:
+            self._const_state[name] = value
+        elif name in JOB_INPUT_ATTRS:
+            self._const_state["job_input"] = self._const_state.get("job_input", {})
+            self._const_state["job_input"][name] = value
+        elif name in NARR_CELL_INFO_ATTRS:
+            self._const_state["job_input"] = self._const_state.get("job_input", {})
+            self._const_state["job_input"]["narrative_cell_info"] = self._const_state["job_input"].get("narrative_cell_info", {})
+            self._const_state["job_input"]["narrative_cell_info"] = value
+        else:
+            object.__setattr__(self, name, value)
+
+    @property
+    def was_terminal(self):
+        """
+        Checks if last queried ee2 state (or those of its children) was terminal.
+        """
+        # add in a check for the case where this is a batch parent job
+        # batch parent jobs with where all children have status "completed" are in a terminal state
+        # otherwise, child jobs may be retried
+        if self._const_state.get("batch_job"):
+            for child_job in self.children:
+                if child_job._last_state and child_job._last_state.get("status") != COMPLETED_STATUS:
+                    return False
+            return True
+
+        else:
+            return (
+                self._last_state
+                and self._last_state.get("status") in TERMINAL_STATUSES
+            )
+
     @property
     def final_state(self):
-        if self.terminal_state is True:
-            return self._last_state
+        if self.was_terminal is True:
+            return self.state()
```

Comment:
Changed from `return self._last_state` to `self.state()` because I implemented `_last_state` as keeping all the attributes and `state()` as doing some of the attribute pruning. Seemed easier to remember that way.